lib/P6doc.pm6

```diff
@@ -1,187 +1,356 @@
+#| Perl6::Documentable objects
+sub process-type-pods(IO::Path @files) returns Array[Perl6::Documentable] is export {
+    my Perl6::Documentable @results;
+
+    for @files.list -> $f {
+        my $documentable = process-pod-source(
+            kind => "type",
+            pod => load($f)[0],
+            filename => $f.Str,
+        );
+        @results.push($documentable);
+    }
+
+    return @results;
+}
+
+#| Search for relevant files in a given directory (recursively, if necessary),
+#| and return a list of the results.
+#| $type-name is the name of the type in the form `Str`, `IO::Spec::Unix` etc..
+#| This assumes that $dir is the base directory for the pod files, example: for
+#| the standard documentation folder 'doc', `$dir` should be `'doc'.IO.add('Type')`.
+sub type-list-files(Str $type-name, $dir) returns Array[IO::Path] is export {
+    my IO::Path @results;
+    my $search-name;
+
+    # TODO: Depth not yet optimal
+    my $query-depth = 0..2;
+
+    my $finder = Path::Finder;
+
+    if $type-name.contains('::') {
+        $query-depth = $type-name.split('::').elems + 1;
+        $search-name = $type-name.split('::').tail;
+
+    } else {
+        $search-name = $type-name;
+    }
+
+    $finder = $finder.depth($query-depth);
+    $finder = $finder.name("{$search-name}.pod6");
+
+    for $finder.in($dir, :file) -> $file {
+        @results.push($file);
+    }
+
+    return @results;
+}
+
+#| Search for a single Routine/Method/Subroutine, e.g. `split`
+# TODO: type-search already profits from pre sieving, routine search does
+# not have any optimization yet!
+sub routine-search(
+    Str $routine,
+    :@topdirs = get-doc-locations() ) returns Array[Perl6::Documentable] is export
+{
+    my Perl6::Documentable @results;
+
+    for @topdirs -> $td {
+        my $registry = compose-registry($td);
+
+        # The result from `.lookup` is containerized, thus we use `.list`
+        for $registry.lookup($routine, :by<name>).list -> $r {
+            @results.append: $r;
+        }
+    }
+
+    return @results
+}
+
+#| Lookup documentation in association with a type, e.g. `Map`, `Map.new`.
+#| `$dir` makes the same assumption as `type-list-files`.
+# TODO: might be cleaner to just give it a list of Perl6::Documentable objects
+sub type-search(
+    Str $type-name,
+    Str :$routine?,
+    IO::Path :$dir where *.d = get-doc-locations().first) returns Array[Perl6::Documentable] is export
+{
+    my Perl6::Documentable @results;
+
+    my IO::Path @files = type-list-files($type-name, $dir);
+    @results = process-type-pods(@files);
+
+    # Old, filter for class
+    #@results = @results.grep: *.subkinds.contains('class');
+    # New, filter only by name, since we already pre sieve with type-list-files
+    @results = @results.grep: *.name eq $type-name;
+
+    # If we provided a routine to search for, we now look for it inside the found classes
+    # and return an array of those results instead
+    if defined $routine {
```

Comment:
Undefined objects evaluate to `False` in boolean contexts I believe, so I think `defined` is redundant here (although I'm not 100% sure).

Reply:
Hmm I actually think I'll leave this one to make it clearer at first glance and remove the (in that case) redundant comment in the future.